crypto/threads_win.c

```diff
@@ -15,6 +15,14 @@
 
 #if defined(OPENSSL_THREADS) && !defined(CRYPTO_TDEBUG) && defined(OPENSSL_SYS_WINDOWS)
 
+# ifdef OPENSSL_USE_APPLINK
```

Comment:
This is not reliable way to determine that target is DLL. There are targets that don't engage it [in shared build], mingw64 and ARM. Applink per se is something one wants to get rid of. It was introduced as assistant for legacy applications, and one really shouldn't encourage its usage. One might also recognize the possibility that OpenSSL can be wrapped in another shared library (when you take static OpenSSL library and create own DLLs without exposing OpenSSL symbols). In this case it would be wrapper that would have to provide dllmain, and it would be only appropriate if OPENSSL_thread_stop() worked. I mean it would be the only thing wrapper would have to call in DLL_THREAD_DETACH case. In other words I'd rather argue in favour of making OPENSSL_thread_stop() work as it stands...

Reply:
There is currently no better way to determine if the target is a DLL.
There should be a way to find out if dllmain.c is linked in.
But I don't see a way how OPENSSL_thread_stop can work without a
race condition or a deadlock of some kind.